docs/dev/style.md

```diff
@@ -925,6 +925,31 @@ Don't use the `ref` keyword.
 Today, it is redundant.
 Between `ref` and mach ergonomics, the latter is more ergonomic in most cases, and is simpler (does not require a keyword).
 
+## Functional Combinators
+
+Use high order monadic combinators like `map`, `then`, only when they are a natural choice, don't bend the code to fit into some combinator.
+If writing a chain of combinators creates friction, replace them with control flow constructs: `for`, `if`, `match`.
+Mostly avoid `bool::then` and `Option::filter`.
+
+```rust
+// GOOD
+if !x.cond() {
+    return None;
+}
+Some(x)
+
+// BAD
+Some(x).filter(|it| it.cond())
+```
+
+This rule is more "soft" then others, and boils down mostly to taste.
+The guiding principle behind this rule is that code should be dense in computation, and sparse in the number of expressions per line.
+The second example contains *less* computation -- `filter` function is an indirection for `if`, it doesn't do any useful work by itself.
+At the same time, it is more crowded -- it takes more time to visually scan it.
```

Comment:
cc @Veykril , @flodiebold , @jonas-schievink , curious what you think about this one!

Reply:
I agree. Combinators are usually overrated, it's what a `for` loop does that's complicated, not the iteration in itself.

Reply:
I agree

Reply:
Although I agree with most changes in #9836 except for the `fold` one; e.g. I wouldn't write a `for` loop that can be directly expressed with `any`.

Reply:
That's a good observation, reworded a bit: https://github.com/rust-analyzer/rust-analyzer/pull/9840/files#diff-80a3e55277ce4f5286b3c7556843e3e9894c23a0c4ba46400372df7fb8b6de24R930

Reply:
Sounds good to me 👍

Reply:
This seems good, yeah